pkg/storage/utils/eosfs/eosfs.go

```diff
@@ -537,22 +543,257 @@ func (fs *eosfs) UnsetArbitraryMetadata(ctx context.Context, ref *provider.Refer
+	}
+	auth, err := fs.getUserAuth(ctx, user, path)
+	if err != nil {
+		return errors.Wrap(err, "eosfs: error getting uid and gid for user")
+	}
+
+	// the cs3apis require to have the write permission on the resource
+	// to set a lock. because in eos we can set attrs even if the user does
+	// not have the write permission, we need to check if the user that made
+	// the request has it
+	has, err := fs.hasWriteAccess(ctx, user, ref)
+	if err != nil {
+		return errors.Wrap(err, "eosfs: cannot check if user has write access on resource")
+	}
+	if !has {
+		return errtypes.BadRequest("user has not write access on resource")
+	}
+
+	encodedLock, err := encodeLock(l)
+	if err != nil {
+		return errors.Wrap(err, "eosfs: error encoding lock")
+	}
+
+	attr := &eosclient.Attribute{
+		Type: UserAttr,
+		Key:  LockKeyAttr,
+		Val:  encodedLock,
+	}
+
+	err = fs.c.SetAttr(ctx, auth, attr, true, false, path)
+	switch {
+	case errors.Is(err, eosclient.AttrAlreadyExistsError):
+		return errtypes.BadRequest("lock already set")
+	default:
+		return err
+	}
+}
+
+func (fs *eosfs) hasWriteAccess(ctx context.Context, user *userpb.User, ref *provider.Reference) (bool, error) {
+	resInfo, err := fs.GetMD(ctx, ref, nil)
+	if err != nil {
+		return false, err
+	}
+	return resInfo.PermissionSet.InitiateFileUpload, nil
+}
+
+func (fs *eosfs) hasReadAccess(ctx context.Context, user *userpb.User, ref *provider.Reference) (bool, error) {
+	resInfo, err := fs.GetMD(ctx, ref, nil)
+	if err != nil {
+		return false, err
+	}
+	return resInfo.PermissionSet.InitiateFileDownload, nil
+}
+
+func encodeLock(l *provider.Lock) (string, error) {
+	data, err := json.Marshal(l)
+	if err != nil {
+		return "", err
+	}
+	return b64.StdEncoding.EncodeToString(data), nil
+}
+
+func decodeLock(raw string) (*provider.Lock, error) {
+	data, err := b64.StdEncoding.DecodeString(raw)
+	if err != nil {
+		return nil, err
+	}
+	l := new(provider.Lock)
+	err = json.Unmarshal(data, l)
+	if err != nil {
+		return nil, err
+	}
+	return l, nil
 }
 
 // RefreshLock refreshes an existing lock on the given reference
-func (fs *eosfs) RefreshLock(ctx context.Context, ref *provider.Reference, lock *provider.Lock) error {
-	return errtypes.NotSupported("unimplemented")
+func (fs *eosfs) RefreshLock(ctx context.Context, ref *provider.Reference, newLock *provider.Lock) error {
+	// TODO (gdelmont): check if the new lock is already expired?
+
+	if newLock.Type == provider.LockType_LOCK_TYPE_SHARED {
+		return errtypes.NotSupported("shared lock not yet implemented")
+	}
+
+	oldLock, err := fs.GetLock(ctx, ref)
+	if err != nil {
+		switch err.(type) {
+		case errtypes.NotFound:
+			// the lock does not exist
+			return errtypes.BadRequest("file was not locked")
+		default:
+			return err
+		}
+	}
+
+	user, err := getUser(ctx)
+	if err != nil {
+		return errors.Wrap(err, "eosfs: error getting user")
+	}
+
+	// check if the user hold the old lock
+	if oldLock.User != nil && !utils.UserEqual(oldLock.User, user.Id) {
```

Comment:
Given that you have the `sameHolder()` check just below, this is redundant actually